homeassistant/components/enphase_envoy/diagnostics.py

```diff
@@ -36,11 +42,81 @@ async def async_get_config_entry_diagnostics(
     """Return diagnostics for a config entry."""
     coordinator: EnphaseUpdateCoordinator = hass.data[DOMAIN][entry.entry_id]
 
-    return async_redact_data(
-        {
-            "entry": entry.as_dict(),
-            "envoy_firmware": coordinator.envoy.firmware,
-            "data": coordinator.data,
-        },
-        TO_REDACT,
+    if TYPE_CHECKING:
+        assert coordinator.envoy.data
+    Envoy_Data: EnvoyData = coordinator.envoy.data
+
+    device_registry = dr.async_get(hass)
+    entity_registry = er.async_get(hass)
+
+    devices = []
+    # for each device associated with the envoy get entity and state information
+    for device in dr.async_entries_for_config_entry(device_registry, entry.entry_id):
+        entities = []
+        for entity in er.async_entries_for_device(
+            entity_registry, device_id=device.id, include_disabled_entities=True
+        ):
+            state_dict = None
+            if state := hass.states.get(entity.entity_id):
+                state_dict = dict(state.as_dict())
+                state_dict.pop("context", None)
+            entities.append({"entity": asdict(entity), "state": state_dict})
+        devices.append({"device": asdict(device), "entities": entities})
+
+    # redact envoy serial in entity data
+    device_entities = json.loads(
+        json.dumps(devices).replace(coordinator.envoy_serial_number, "<<envoyserial>>")
+    )
+
+    # redact envoy serial in envoy data
+    cleaned_data = json.loads(
+        json.dumps(coordinator.data).replace(
+            coordinator.envoy_serial_number, "<<envoyserial>>"
+        )
     )
```

Comment:
You know we have the redact data for this purpose right?

Reply:
Yes, but that replace whole fields and not specified text in a field.